### PR TITLE
feat: [CI-10273] Added the support of dynamic timeout for keys in redis stream

### PR DESCRIPTION
**Pull request by ShobhitSingh11**
## 
Initially the timeout for write in redis stream was 5 hours and it was constant, now it is changed to dynamic. We can still download content from blob even after 5 hours depends upon what's the timeout being set. However maximum timeout being supported is 24 Hours. Tested with postman.

Here is the API we are using => https://localhost:8181/log-service/stream?accountID=kmpySmUISimoRrJL6NL73w&key=kmpySmUISimoRrJL6NL73w%2Fpipeline%2FCI10273%2F3%2F-4NYs6YQOTQu-Q4XpP6JarQ%2FApproval%2FCustom_Approval-commandUnit%3AExecute&timeout_seconds=15

We've added new query parameter "timeout_seconds" at the end.

Here are the cases we have handled and tested
Case: 1 => When timeout provided is negative or in invalid format.

https://github.com/harness/harness-core/assets/139750384/9365c2fa-3945-49de-abf9-7a5e5b08bb07

Case: 2 => We are providing timeout parameter. In video we can see the key starting with log-servicekmpy.... which will be disappeared after 25 seconds of creating the stream.

https://github.com/harness/harness-core/assets/139750384/b30b651b-d665-4947-8188-378373f173bf

Case: 3 => Here we are not providing any timeout parameter, so it will take default value. For testing purpose we have set it to 40 seconds but it is actually 5 hours. We will see the key starting with log-servicekmpy.... will disappear after 40 seconds from stream creation.
 

https://github.com/harness/harness-core/assets/139750384/251a48a2-8f40-4385-a7da-4ffd4d26824f

Here is the error message when Key is not present still we trying to write to the same key (Post timeout)

Screenshot 2023-12-20 at 12 00 36 PM

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

### Rebase Instructions:
To rebase your branch onto the latest changes in the target branch, simply use the following command in a comment: `/rebase`

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/56923)

### CODE REVIEW

**Comment by ShobhitSingh11** [comment]
trigger smartchecks

**Comment by ShobhitSingh11** [comment]
trigger uts

**Comment by ShobhitSingh11** [comment]
trigger gobuild

**Comment by ShobhitSingh11** [comment]
trigger smartchecks

**Comment by ShobhitSingh11** [comment]
trigger uts

**Review by ShobhitSingh11 — COMMENTED** [review]
Please ignore this file, will revert after final review. This was unversioned file.

File: .vscode/launch.json
Comment (ShobhitSingh11) [inline .vscode/launch.json:1]: Please ignore this file. Will revert after final review. It was unversioned and got committed

**Comment by ShobhitSingh11** [comment]
trigger uts

**Comment by ShobhitSingh11** [comment]
trigger ut0

**Comment by ShobhitSingh11** [comment]
trigger smartchecks

**Comment by ShobhitSingh11** [comment]
trigger uts

**Comment by ShobhitSingh11** [comment]
trigger gobuild

File: product/log-service/handler/stream.go [hunk 1]
```diff
@@ -48,7 +55,32 @@ func HandleOpen(stream stream.Stream) http.HandlerFunc {
...
+		var keyExpiryTimeSeconds int
+
+		if keyExpiryTimeSecondsString == "" {
+			keyExpiryTimeSeconds = defaultKeyExpiryTimeSeconds
+		} else {
+			var err error
+			keyExpiryTimeSeconds, err = strconv.Atoi(keyExpiryTimeSecondsString)
+
+			if err != nil {
+				WriteBadRequest(w, errors.New("Key expiry timeout is in invalid format, it should be in integer"))
```
Comment (vistaarjuneja) [inline product/log-service/handler/stream.go:68]: Wrap the error from Atoi using `fmt.Errorf("unable to parse timeout_secs: %w", err)`

File: product/log-service/handler/stream.go [hunk 2]
```diff
@@ -48,7 +55,32 @@ func HandleOpen(stream stream.Stream) http.HandlerFunc {
...
+				return
+			}
+
+			if keyExpiryTimeSeconds < 0 {
+				WriteBadRequest(w, errors.New("Key expiry timeout cannot be negative"))
```
Comment (vistaarjuneja) [inline product/log-service/handler/stream.go:73]: Keep error messages that we return to user lower case everywhere

File: product/log-service/handler/stream.go [hunk 3]
```diff
@@ -194,6 +226,7 @@ func HandleWrite(s stream.Stream, metrics *metric.Metrics) http.HandlerFunc {
 
 		// write to stream only if it exists
 		if err := s.Exists(ctx, key); err != nil {
+            WriteNotFound(w, errors.New("Key not found. Either key is deleted or does not exist"))
```
Comment (smjt-h) [inline product/log-service/handler/stream.go:229]: What does it return currently if we try to write to a stream that is expired

File: .vscode/launch.json [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+    "configurations": [
+        {
+            "name": "Launch Package",
+            "type": "go",
+            "request": "launch",
+            "mode": "auto",
+            "program": "${fileDirname}"
+        }
+    ]
+}
```
Comment (smjt-h) [inline .vscode/launch.json:15]: you can remove

File: product/log-service/handler/stream.go (hunk 1 above)
Comment (smjt-h) [inline product/log-service/handler/stream.go:68]: keep message in lower case

File: product/log-service/stream/redis/redis.go [hunk 5]
```diff
@@ -93,7 +93,7 @@ func (r *Redis) Create(ctx context.Context, key string) error {
 		return errors.Wrap(err, fmt.Sprintf("could not create stream with key: %s", key))
 	}
 
-	r.setExpiry(key, defaultKeyExpiryTimeSeconds*time.Second)
+	r.setExpiry(key, time.Duration(keyExpiryTimeSeconds)*time.Second)
```
[the remark this answers is not in the document]
Comment (ShobhitSingh11) [reply product/log-service/stream/redis/redis.go:96]: Here parameter type for setExpiry method is time.Duration

File: product/log-service/handler/stream.go (hunk 3 above)
Comment (ShobhitSingh11) [reply product/log-service/handler/stream.go:229]: It is being removed

**Comment by ShobhitSingh11** [comment]
trigger smartchecks

**Comment by ShobhitSingh11** [comment]
trigger uts

**Review by smjt-h — DISMISSED** [review]

File: product/log-service/stream/stream.go [hunk 6]
```diff
@@ -22,7 +22,7 @@ const Prefix = "log-service"
 // Stream defines the live log streaming interface.
 type Stream interface {
 	// Create creates the log stream for the string key.
-	Create(context.Context, string) error
+	Create(context.Context, string, int) error
```
Comment (vistaarjuneja) [inline product/log-service/stream/stream.go:25]: Nit: let's name parameters now that we have something not super intuitive.
`Create(ctx context.Context, key string, timeoutSecs int)`

**Comment by vistaarjuneja** [comment]
Nice work @ShobhitSingh11, I like all the screen recordings for various cases!

**Review by vistaarjuneja — DISMISSED** [review]

**Comment by ShobhitSingh11** [comment]
trigger smartchecks

**Comment by ShobhitSingh11** [comment]
trigger uts

**Comment by ShobhitSingh11** [comment]
trigger gobuild

**Review by vistaarjuneja — APPROVED** [review]

**Review by smjt-h — APPROVED** [review]

**Pull request merged**